Refit/CachedRequestBuilderImplementation.cs

```diff
@@ -0,0 +1,58 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+
+namespace Refit
+{
+    public class CachedRequestBuilderImplementation : IRequestBuilder
+    {
+        readonly IRequestBuilder innerBuilder;
+        readonly ConcurrentDictionary<string, Func<HttpClient, object[], object>> methodDictionary = new ConcurrentDictionary<string, Func<HttpClient, object[], object>>();
+
+        public CachedRequestBuilderImplementation(IRequestBuilder innerBuilder)
+        {
+            this.innerBuilder = innerBuilder;
+        }
+
+        public Func<HttpClient, object[], object> BuildRestResultFuncForMethod(string methodName, Type[] parameterTypes = null, Type[] genericArgumentTypes = null)
+        {
+            string cacheKey = GetCacheKey(methodName, parameterTypes, genericArgumentTypes);
+            var func = methodDictionary.GetOrAdd(cacheKey, _ => innerBuilder.BuildRestResultFuncForMethod(methodName, parameterTypes, genericArgumentTypes));
+
+            return func;
+        }
+
+        string GetCacheKey(string methodName, Type[] parameterTypes, Type[] genericArgumentTypes)
+        {
+            string genericDefinition = GetGenericString(genericArgumentTypes);
+            string argumentString = GetArgumentString(parameterTypes);
+
+            return $"{methodName}{genericDefinition}({argumentString})";
+        }
+
+        string GetArgumentString(Type[] parameterTypes)
+        {
+            if (parameterTypes == null || parameterTypes.Length == 0)
+            {
+                return "";
+            }
+
+            return parameterTypes.Select(t => t.Name).Aggregate((s1, s2) => s1 + ", " + s2);
```

Comment:
`string.Join(", ", parameterTypes.Select(t => t.Name)` would be a little clearer here. There's another one like this in the next method too.